File: lengthOfLongestSubstring.py
#sliding window/two pointer
#time complexity -> O(n) total number of char in the string
#space complexity O(m) total number of unique char in string
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        res = 0
        charSet = set()

        #r in this case will represent the right pointer
        for r in range(len(s)):
            #if r is a duplicate from discovered
            while s[r] in charSet:
                #move the left pointer until right pointer char is no longer a dup
                charSet.remove(s[l])
                l += 1
            charSet.add(s[r])
            res = max(res, r-l+1)

        return res

File: lengthOfLongestSubstring.cpp
//sliding windown
//time complexity -> O(n) goes through all char in s with for loop
//space complexity -> O(m) unique characters stored in set
#include <unordered_set>
//using namespace std;
class Solution{
public:
    int lengthOfLongestSubstring(string s){
        int l = 0;
        std::unordered_set<char> charSet;
        int res = 0;

        for(int r = 0; r < s.length(); r++){
            while(charSet.find(s[r]) != charSet.end()){
                charSet.erase(s[l]);
                l++;
            }
            charSet.insert(s[r]);
            res = max(res, r-l+1);
        }
        return res;
    }
};